From b7b5876bc7dbbcc663719f1c6a5ee6496b092f0f Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Fri, 28 Aug 2020 13:51:47 +0200
Subject: [PATCH 1/1] efi_loader: correct SizeOfImage for UEFI binaries

Linking with objcopy does not set SizeOfImage and we do not know the file
size at compile time.

Run the efifixup tool on all generated UEFI binaries to set the file size.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 Makefile             | 3 ++-
 scripts/Makefile.lib | 3 +++
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 5dd4c6bd40..7bd0199e08 100644
--- a/Makefile
+++ b/Makefile
@@ -412,6 +412,7 @@ PYTHON2		= python2
 PYTHON3		= python3
 DTC		?= $(objtree)/scripts/dtc/dtc
 CHECK		= sparse
+EFIFIXUP	?= $(objtree)/tools/efifixup
 
 CHECKFLAGS     := -D__linux__ -Dlinux -D__STDC__ -Dunix -D__unix__ \
 		  -Wbitwise -Wno-return-void -D__CHECK_ENDIAN__ $(CF)
@@ -454,7 +455,7 @@ export ARCH CPU BOARD VENDOR SOC CPUDIR BOARDDIR
 export CONFIG_SHELL HOSTCC KBUILD_HOSTCFLAGS CROSS_COMPILE AS LD CC
 export CPP AR NM LDR STRIP OBJCOPY OBJDUMP KBUILD_HOSTLDFLAGS KBUILD_HOSTLDLIBS
 export MAKE LEX YACC AWK PERL PYTHON PYTHON2 PYTHON3
-export HOSTCXX KBUILD_HOSTCXXFLAGS CHECK CHECKFLAGS DTC DTC_FLAGS
+export HOSTCXX KBUILD_HOSTCXXFLAGS CHECK CHECKFLAGS DTC DTC_FLAGS EFIFIXUP
 
 export KBUILD_CPPFLAGS NOSTDINC_FLAGS UBOOTINCLUDE OBJCOPYFLAGS KBUILD_LDFLAGS
 export KBUILD_CFLAGS KBUILD_AFLAGS
diff --git a/scripts/Makefile.lib b/scripts/Makefile.lib
index 56e9d54242..976f111970 100644
--- a/scripts/Makefile.lib
+++ b/scripts/Makefile.lib
@@ -399,8 +399,11 @@ cmd_efi_objcopy = $(OBJCOPY) -j .header -j .text -j .sdata -j .data -j \
 		.dynamic -j .dynsym  -j .rel* -j .rela* -j .reloc \
 		$(if $(EFI_TARGET),$(EFI_TARGET),-O binary) $^ $@
 
+quiet_cmd_efi_fixup = EFIFIXUP $@
+cmd_efi_fixup = $(EFIFIXUP) $@
 $(obj)/%.efi: $(obj)/%_efi.so
 	$(call cmd,efi_objcopy)
+	$(call cmd,efi_fixup)
 
 quiet_cmd_efi_ld = LD      $@
 cmd_efi_ld = $(LD) -nostdlib -znocombreloc -T $(EFI_LDS_PATH) -shared \
-- 
2.28.0

